File: HashMapJogo.java
import java.util.*;

class Jogo {
    private Map<String, Integer> jogadores = new HashMap<>();

    public void adicionarJogador(String nome, Integer pontos) {
        jogadores.put(nome, pontos);
    }

    public void removerJogador(String nome) {
        jogadores.remove(nome);
    }

    public void listarJogadores() {
        jogadores.entrySet()
            .stream()
            .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
            .forEach(entry -> 
                System.out.println("Jogador: " + entry.getKey() + ", pontos: " + entry.getValue()));
    }

    public void atualizarPontuacaoJogador(String nomeJogador, Integer novaPontuacao) {
        if (jogadores.containsKey(nomeJogador)) {
            jogadores.put(nomeJogador, novaPontuacao);
        } else {
            System.out.println("Jogador " + nomeJogador + " não encontrado.");
        }
    }

    public void jogadorVencedor() {
        Optional<Map.Entry<String, Integer>> jogadorOpt = jogadores.entrySet()
            .stream()
            .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
            .findFirst();

        if (jogadorOpt.isPresent()) {
            Map.Entry<String, Integer> jogador = jogadorOpt.get();
            System.out.println("Jogador vencedor: " + jogador.getKey() + ", pontos: " + jogador.getValue());
        } else {
            System.out.println("Não há jogadores.");
        }
    }
}

public class HashMapJogo {
    public static void main(String[] args) {
        Jogo jogo = new Jogo();

        jogo.adicionarJogador("Marcos", 100);
        jogo.adicionarJogador("Ingrid", 1);

        jogo.atualizarPontuacaoJogador("Ingrid", 20);

        jogo.listarJogadores();
        jogo.jogadorVencedor();
    }
}

File: HashMapJogo.py
class Jogo:
    def __init__(self):
        self.jogadores = dict()

    def adicionarJogador(self, nome, pontos):
        self.jogadores[nome] = pontos

    def removerJogador(self, nome):
        if nome in self.jogadores:
            del self.jogadores[nome]

    def listarJogadores(self):
        jogadoresOrdenados = sorted(self.jogadores.items(), key=lambda jogador: jogador[1], reverse=True)
        for jogador in jogadoresOrdenados:
            print("Jogador: {}, pontos: {}".format(jogador[0], jogador[1]))

    def atualizarPontuacaoJogador(self, nomeJogador, novaPontuacao):
        if nomeJogador in self.jogadores:
            self.jogadores[nomeJogador] = novaPontuacao
        else:
            print("Jogador {} não encontrado.".format(nomeJogador))

    def jogadorVencedor(self):
        if not self.jogadores:
            print("Não há jogadores.")
            return

        jogadorVencedor = max(self.jogadores.items(), key=lambda jogador: jogador[1])
        print("Jogador vencedor: {}, pontos: {}".format(jogadorVencedor[0], jogadorVencedor[1]))


jogo = Jogo()
jogo.adicionarJogador("Marcos", 100)
jogo.adicionarJogador("Ingrid", 1)
jogo.atualizarPontuacaoJogador("Ingrid", 20)
jogo.listarJogadores()
jogo.jogadorVencedor()
